### ISSUE TITLE
Allow for "prepared provider" stations

### ISSUE THREAD

**Issue by NicolBolas** [mamoniot/project-cybersyn#110]
One problem I have with CyberSyn is that it is less responsive than vanilla trains can be. At least, not unless you make room for stackers.

The issue is essentially this: if the round-trip time for a train to pick up a load, go to the requester, and unload is less than the time it takes to *consume* that load, then there's a problem. Now, you might think that this is unfixable, that it's inherent to how a train network would work.

But it is *possible* to solve it. So long as the provider is able to get in new stuff fast enough (or you have multiple providers that in aggregate can keep up), all you need to do is to have a train being loaded at the provider *before* the previous train has finished unloading at the requester.

With vanilla trains, I wouldn't have this problem because every train only serviced a single material. As such, a train can just sit at a provider station and wait with a full load until a requester slot for that item opens up. Once the train leaves that provider, another train can immediately pull in and start loading, even if there's no active request available.

CyberSyn doesn't really have a way to handle this scenario. Well, it can if you have a stacker at the requester and a train limit > 1, but stackers take up space.

What is desired is a way to set a provider station to only ever field a single request. If the provider has enough materials for that request, even if nobody is actually requesting it yet, then the provider will call for a (currently unused) train to load up that request. Then the train will sit there. The usual logic for deciding which station should be used to field a request will be used, but if a prepared provider is selected, the current train carrying the stuff will field it.

And once that train leaves, if the provider has enough materials to service its prepared request, then it will call for another train to prepare for a future request.

Note that prepared providers should not have special priority over non-prepared providers. If the user wants to give them priority, there's already a priority system in place that can handle this. It can even be dynamically set based on whether there is a train currently sitting at the station.

Note that this would make it easier to use CyberSyn with direct mining setups, where train loading times can be substantial (until you get into very high productivity).

**Comment by asvanberg** [comment]
> The issue is essentially this: if the round-trip time for a train to pick up a load, go to the requester, and unload is less than the time it takes to _consume_ that load, then there's a problem. Now, you might think that this is unfixable, that it's inherent to how a train network would work.

You solve this by making a bigger request at your requesting station. If it consumes say 100 items/s and a train carries 8000 items with a round-trip time of two minutes. Then you know that in the time it takes for a request to be delivered it will consume 100*120=12000 items, that means your request has to be 12000+8000=20000 items. Once the station is down to 12000 a train will be dispatched for 8000 and it can eat through the buffer while waiting. After 80 seconds the train has not yet arrived but now the buffer is down to 4000 items and so a second train will be dispatched.

As long as your request is big enough then train will be dispatched every 80 seconds to keep up with the consumption rate.

**Comment by NicolBolas** [comment]
> After 80 seconds the train has not yet arrived but now the buffer is down to 4000 items and so a second train will be dispatched.

And there is the problem. A second train can only be dispatched if the train limit is higher than 1. And that requires that there's a stacker at the requester. The entire point of my idea is to *not* have a stacker there (because those take up space and I want to be able to take up less space).

**Comment by asvanberg** [comment]
> > After 80 seconds the train has not yet arrived but now the buffer is down to 4000 items and so a second train will be dispatched.
> 
> And there is the problem. A second train can only be dispatched if the train limit is higher than 1. And that requires that there's a stacker at the requester. The entire point of my idea is to _not_ have a stacker there (because those take up space and I want to be able to take up less space).

No, you do not need a stacker in this case. The first train will have unloaded and left before the second train arrives.
